// @puzzlet/templatedx/dist/filter-plugins/index.js
export const capitalize = (input) => {
    if (typeof input !== "string")
        return input;
    return input.charAt(0).toUpperCase() + input.slice(1);
};
export const upper = (input) => {
    if (typeof input !== "string")
        return input;
    return input.toUpperCase();
};
export const lower = (input) => {
    if (typeof input !== "string")
        return input;
    return input.toLowerCase();
};
export const truncate = (input, length) => {
    if (typeof input !== "string")
        return input;
    if (input.length <= length)
        return input;
    return input.substring(0, length) + "...";
};
export const abs = (input) => {
    return Math.abs(input);
};
export const join = (input, separator = ", ") => {
    if (!Array.isArray(input))
        return input;
    return input.join(separator);
};
export const round = (input, decimals = 0) => {
    return Number(Math.round(Number(input + "e" + decimals)) + "e-" + decimals);
};
export const replace = (input, search, replace) => {
    if (typeof input !== "string")
        return input;
    return input.split(search).join(replace);
};
export const urlencode = (input) => {
    if (typeof input !== "string")
        return input;
    return encodeURIComponent(input);
};
export const dump = (input) => {
    return JSON.stringify(input);
};
